apps/aecore/lib/aecore/chain/chainstate.ex

```diff
@@ -106,23 +106,51 @@ defmodule Aecore.Chain.Chainstate do
     }
   end
 
-  defp calculate_miner_reward_chain_state(txs, chainstate, block_height, miner) do
-    case miner do
-      @genesis_miner ->
-        chainstate
+  defp calculate_miner_reward_chain_state(chainstate, block_height) do
```

Comment:
If we have a fork like:
```
           [Gen2a] - ... - [Gen#{GovernanceConstants.beneficiary_reward_lock_time()+1}a] 
           //
[Gen1]--
            \\
            [Gen2b] - ... - [Gen#{GovernanceConstants.beneficiary_reward_lock_time()+1}b]
```
The reward calculation will apply wrong rewards on one of the branches. This situation could happen if a group of miners decides to build blocks on top of the mallicoius chain instead of building on top the most difficult chain. This could lead to miscalculations in the main chain which could invalidate the the main chain and lead to double spend attacks.

Make sure on what fork the block reward was calculated.

Reply:
I know, we need to handle forks generally, not only when calculating rewards. Currently the check for new blocks in add_validated_block is if top_height <= new_block_height, we should be comparing difficulties not height. Haven't thought about a way but I won't do it here...